src/meta/src/hummock/manager/compaction.rs

```diff
@@ -117,76 +121,186 @@ impl HummockManager {
+                    compact_ret,
+                    compactor.clone(),
+                    no_task_groups,
+                    existed_groups,
+                );
+
+                // if no task or failed task, break
+                if picked_tasks_count == 0 || !failed_tasks.is_empty() {
+                    break;
+                }
+
+                generated_task_count += picked_tasks_count;
+            }
+
+            (failed_tasks, generated_task_count)
+        }
+
         if let Some(compactor) = self.compactor_manager.get_compactor(context_id) {
             let (groups, task_type) = self.auto_pick_compaction_groups_and_type().await;
+
+            let mut generated_task_count = 0;
+            let mut existed_groups = groups.clone();
+            let mut no_task_groups: HashSet<CompactionGroupId> = HashSet::default();
+            let mut failed_tasks = vec![];
             if !groups.is_empty() {
+                // batch get multi groups
                 let selector: &mut Box<dyn CompactionSelector> =
                     compaction_selectors.get_mut(&task_type).unwrap();
+                let (tmp_failed_tasks, picked_tasks_count) = handle_pick_tasks(
+                    self,
+                    pull_task_count,
+                    groups.clone(),
+                    compactor.clone(),
+                    selector,
+                    &mut no_task_groups,
+                    &mut existed_groups,
+                )
+                .await;
+
+                failed_tasks.extend(tmp_failed_tasks);
+                generated_task_count += picked_tasks_count;
+            }
 
-                let mut generated_task_count = 0;
-                let mut existed_groups = groups.clone();
-                let mut no_task_groups: HashSet<CompactionGroupId> = HashSet::default();
-                let mut failed_tasks = vec![];
+            // unschedule cg
+            for cg_id in no_task_groups {
+                if failed_tasks.is_empty() && generated_task_count < pull_task_count {
+                    let versioning_guard = read_lock!(self, versioning).await;
+                    let versioning = versioning_guard.deref();
 
-                while generated_task_count < pull_task_count && failed_tasks.is_empty() {
-                    let compact_ret = self
-                        .get_compact_tasks(
-                            existed_groups.clone(),
+                    let write_stop = versioning.write_limit.contains_key(&cg_id);
+                    let enable_emergency_picker = match self
+                        .compaction_group_manager
+                        .read()
+                        .await
+                        .try_get_compaction_group_config(cg_id)
+                    {
+                        Some(config) => config.compaction_config.enable_emergency_picker,
+                        None => {
+                            unreachable!("compaction-group {} not exist", cg_id)
```

Comment:
Is it really unreachable? Will compaction group be destroyed if the table in the cg is dropped?